Source/Core/CameraLimiter.js

```diff
@@ -0,0 +1,301 @@
+        var returningPosition = position;
+
+        returningPosition.x = Math.max(axisAlignedMinimum.x, Math.min(axisAlignedMaximum.x, returningPosition.x));
+        returningPosition.y = Math.max(axisAlignedMinimum.y, Math.min(axisAlignedMaximum.y, returningPosition.y));
+        returningPosition.z = Math.max(axisAlignedMinimum.z, Math.min(axisAlignedMaximum.z, returningPosition.z));
+
+        return returningPosition;
+    };
+
+    /**
+     * @private
+     */
+    CameraLimiter.prototype._closestLocationInBoundingRectangle = function(position) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('this.boundingObject.x', this.boundingObject.x);
+        Check.defined('this.boundingObject.y', this.boundingObject.y);
+        Check.defined('this.boundingObject.width', this.boundingObject.width);
+        Check.defined('this.boundingObject.height', this.boundingObject.height);
+        //>>includeEnd('debug');
+
+        var rect = this.boundingObject;
+
+        var rectMinimum = new Cartesian2(rect.x, rect.y);
+        var rectMaximum = new Cartesian2(rect.x + rect.width, rect.y + rect.height);
+        var returningPosition = new Cartesian2(position.x, position.y);
+
+        returningPosition.x = Math.max(rectMinimum.x, Math.min(rectMaximum.x, returningPosition.x));
+        returningPosition.y = Math.max(rectMinimum.y, Math.min(rectMaximum.y, returningPosition.y));
+
+        return new Cartesian3(returningPosition.x, returningPosition.y, position.z);
+    };
+
+    /**
+     * @private
+     */
+    CameraLimiter.prototype._closestLocationInBoundingSphere = function(position) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('this.boundingObject.radius', this.boundingObject.radius);
+        Check.defined('this.boundingObject.center', this.boundingObject.center);
+        //>>includeEnd('debug');
+
+        var center = this.boundingObject.center;
+        var radius = this.boundingObject.radius;
+
+        // to avoid dividing by zero check if same location as center.
+        if (position.equals(center)) {
+            return position;
+        }
+
+        // convert to cartographic and check if at appropriate height - sphere radius - if not then clarify
+        // var cartographicRep = Cartographic.fromCartesian(position);
+        // if (cartographicRep.height > )
+
+        var direction = Cartesian3.subtract(center, position, new Cartesian3());
+        Cartesian3.normalize(direction, direction);
+        Cartesian3.multiplyByScalar(direction, radius, direction);
+        var locationOnSphereInDirection = Cartesian3.add(direction, center, new Cartesian3());
+
+        var distanceToLocOnSphere = Cartesian3.distanceSquared(locationOnSphereInDirection, center);
+        var distanceToPosition = Cartesian3.distanceSquared(position, center);
+
+        return (distanceToPosition < distanceToLocOnSphere || distanceToPosition) ? position : locationOnSphereInDirection;
```

Comment:
Can you find out if it's inside the sphere earlier and return the original position earlier in the function? That would avoid unnecessary operations and make this more readable as well.